### PR TITLE
WIP: add working examples with Jaeger tracer.

### PR DESCRIPTION
**Pull request by ror6ax**
I've added examples that use Jaeger instead of Lightstep, introduce serialization to the wire, and are bare minimal for ease of understanding.

Before running both files you need to set `JAEGER_HOST` and `WEBSERVER_HOST` environment variables.

@bhs  @yurishkuro - please let me know if the approach seems good for you. I'll be writing README after your thumbs up on general direction.

### CODE REVIEW

File: example/client.py [hunk 1]
```diff
@@ -1,66 +1,35 @@
...
+from opentracing.propagation import Format
 
-app = Flask(__name__)
+from os import environ
 
-# one-time tracer initialization code
-ls_tracer = lightstep.tracer.init_tracer(group_name="example client", access_token="{your_lightstep_token}")
-# this tracer does not trace all requests, so the @tracer.trace() decorator must be used
-tracer = FlaskTracer(ls_tracer)
+JAEGER_HOST = environ.get('JAEGER_HOST')
```
Comment (yurishkuro) [inline example/client.py]: maybe default to local host?

File: example/client.py [hunk 2]
```diff
@@ -1,66 +1,35 @@
...
 
-@app.route("/")
-def index():
-	'''
-	Index page, has no tracing.
-	'''
-	return "Index Page"
+WEBSERVER_HOST = environ.get('WEBSERVER_HOST')
 
+config = Config(config={'sampler': {'type': 'const', 'param': 1},
```
Comment (yurishkuro) [inline example/client.py]: this could use a comment explaining what's happening, i.e. "configuring and creating an instance of Jaeger Tracer with 100% sampling and debug logging of spans".

File: example/client.py [hunk 3]
```diff
@@ -1,66 +1,35 @@
...
-		request = urllib2.Request(url)
-		inject_as_headers(ls_tracer, span, request)
-		try:
-			response = urllib2.urlopen(request)
-		except urllib2.URLError as ue:  
-			response = ue
-	for i in range(numrequests):
-		send_request()
-	return "Requests sent"
+install_all_patches()
```
Comment (yurishkuro) [inline example/client.py]: could use a comment explaining what this does

File: example/client.py [hunk 4]
```diff
@@ -1,66 +1,35 @@
...
-	span = tracer.get_span()
-	span.log_event("hello world")
-	return "Something was logged"
+url = "http://{}:5000/log".format(WEBSERVER_HOST)
+with tracer.start_span('say-hello') as span:
+    span.set_tag(tags.HTTP_METHOD, 'GET')
+    span.set_tag(tags.HTTP_URL, url)
+    span.set_tag(tags.SPAN_KIND, tags.SPAN_KIND_RPC_CLIENT)
+    headers = {}
+    tracer.inject(span, Format.HTTP_HEADERS, headers)
```
Comment (yurishkuro) [inline example/client.py]: sorry, why is this needed? Since you're calling `install_all_patches()` above, the `requests` module is automatically instrumented and will do the exact same thing.

File: example/client.py [hunk 5]
```diff
@@ -1,66 +1,35 @@
...
-
-def inject_as_headers(tracer, span, request):
-    text_carrier = {}
-    tracer.inject(span.context, opentracing.Format.TEXT_MAP, text_carrier)
-    for k, v in text_carrier.iteritems():
-        request.add_header(k,v)
+    r = requests.get(url, headers=headers)
+    print(r.text)
 
+    time.sleep(2)
```
Comment (yurishkuro) [inline example/client.py]: this may confuse people, I recommend adding a comment
```
# allow tracer to flush the spans - https://github.com/jaegertracing/jaeger-client-python/issues/50
```

File: example/server.py [hunk 6]
```diff
@@ -1,41 +1,40 @@
...
+        app = Flask(__name__)
+        log_level = logging.DEBUG
+        logging.getLogger('').handlers = []
+        logging.basicConfig(format='%(asctime)s %(message)s', level=log_level)
+        config = Config(config={'sampler': {'type': 'const', 'param': 1},
+                                'logging': True,
+                                'local_agent':
+                                {'reporting_host': JAEGER_HOST}},
+                        service_name="jaeger_opentracing_example")
+        ls_tracer = config.initialize_tracer()
```
Comment (yurishkuro) [inline example/server.py]: rename to `jaeger_tracer`

File: example/server.py [hunk 7]
```diff
@@ -1,41 +1,40 @@
...
+        tracer = FlaskTracer(ls_tracer)
+
+        @app.route('/log')
+        @tracer.trace()
+        def log():
+
+                span_ctx = ls_tracer.extract(Format.HTTP_HEADERS, request.headers)
```
Comment (yurishkuro) [inline example/server.py]: similar to client, the point of using `@tracer.trace()` is that the handlers are automatically instrumented, but you're doing it again manually

File: example/README.md [hunk 8]
```diff
@@ -1,74 +1,81 @@
...
+```
+> python3 client.py   
+```
 
-Navigate to `/request/simple/<int:numrequests>` where numrequests is the number
-of requests you want to send to the page. This will send a request to the server
-app, and the trace will include a span on both the client and server sides. This
-occurs automatically since both the client and server functions are decorated
-with @tracer.trace().
+To see the traced requests, go to Jaeger web interface.
```
Comment (yurishkuro) [inline example/README.md]: "and refresh the page"

File: example/README.md [hunk 9]
```diff
@@ -1,74 +1,81 @@
...
-span = tracer.get_span()
-span.log_event("hello world")
-```
-Result for `/log`:
+### Trace a request from browser:
 
-![span with log](https://raw.githubusercontent.com/kcamenzind/flask_opentracing/master/example/img/log.png)
+Browse to http://localhost:5000/log and compare the trace in Jaeger.
+The last one has 2 spans instead of 3. The span of webserver's GET method is missing.
+That is because in client.py we pass already existing span over the wire, whereas the request from your browser has to tracing data in it.
```
Comment (yurishkuro) [inline example/README.md]: That is because `client.py` starts a trace and passes trace context over the wire, whereas the request from your browser has to tracing context in it.

File: example/README.md [hunk 10]
```diff
@@ -1,74 +1,81 @@
...
 
 ### Add spans to the trace manually:
 
-Navigate to `/request/childspan/<int:numrequests>`. This will send a request to
-the server app, and the trace will include a span on both the client and server
-sides. The server app will additionally create a child span for the server-side
-span. This example demonstrates how you can trace non-RPC function calls in your
-app, through accessing the current span as follows:
+In log function of the server app, we are creating current_span. This is done to
+trace the work that is being done to render the response to /log. Suppose there's
```
Comment (yurishkuro) [inline example/README.md]: "response to `/log` endpoint"

File: example/README.md [hunk 11]
```diff
@@ -1,74 +1,81 @@
...
+a database connection happening. By creating a separate span for it, you'll be able
+to trace the DB request separately from rendering or the response. This gives a
+lot of flexibility to the user.
```
Comment (yurishkuro) [inline example/README.md]: maybe it's worth mentioning that `install_patches` will do that automatically for some frameworks like SqlAlchemy, Redis client, etc.

File: example/server.py [hunk 12]
```diff
@@ -1,41 +1,43 @@
...
+                                'local_agent':
+                                # Also, provide a hostname of Jaeger instance to send traces to.
+                                {'reporting_host': JAEGER_HOST}},
+                        # Service name can be arbitrary string describing this particular web service.
+                        service_name="jaeger_opentracing_example")
+        jaeger_tracer = config.initialize_tracer()
+        tracer = FlaskTracer(jaeger_tracer)
+
+        @app.route('/log')
+        # Indicate that /log endpoint should be traced
```
Comment (yurishkuro) [inline example/server.py]: nit: this would read better on the same line as the decorator

File: example/server.py [hunk 13]
```diff
@@ -1,41 +1,43 @@
...
+        @tracer.trace()
+        def log():
+                # Extract the span information for request object.
+                request_span = tracer.get_span(request)
+                # Start internal span to trace some kind of business logic for /log.
+                current_span = jaeger_tracer.start_span("python webserver internal span of log method",
```
Comment (yurishkuro) [inline example/server.py]: the preferred syntax here would be 
```python
with jaeger_tracer.start_span("inner-span", child_of=request_span) as span:
    # Perform some computations to be traced.
    span.log_kv({'event': 'my computer knows math!', 'result': c})
```

**Comment by ror6ax** [comment]
@yurishkuro Fixed per your comments.

File: example/README.md [hunk 14]
```diff
@@ -1,74 +1,85 @@
...
+docker run -d -e \
+  COLLECTOR_ZIPKIN_HTTP_PORT=9411 \
+  -p 5775:5775/udp \
+  -p 6831:6831/udp \
+  -p 6832:6832/udp \
+  -p 5778:5778 \
+  -p 16686:16686 \
+  -p 14268:14268 \
+  -p 9411:9411 \
+  jaegertracing/all-in-one:latest
```
Comment (yurishkuro) [inline example/README.md]: would it make sense to just use host networking? Most of these ports are not actually necessary for this example anyway.

File: example/README.md [hunk 15]
```diff
@@ -1,74 +1,85 @@
...
+In log function of the server app, we are creating current_span. This is done to
+trace the work that is being done to render the response to /log endpoint. Suppose there's
+a database connection happening. By creating a separate span for it, you'll be able
+to trace the DB request separately from rendering or the response. This gives a
+lot of flexibility to the user.
 
-### Add spans to the trace manually:
+Speaking about databases, using install_all_patches() method from
+opentracing_instrumentation package gives you a way to trace
+your MySQLdb, SQLAlchemy, Redis queries with without writing boilerplate code.
```
Comment (yurishkuro) [inline example/README.md]: typo: "with without"

File: example/server.py [hunk 16]
```diff
@@ -1,41 +1,41 @@
...
+                        service_name="jaeger_opentracing_example")
+        jaeger_tracer = config.initialize_tracer()
+        tracer = FlaskTracer(jaeger_tracer)
+
+        @app.route('/log')
+        @tracer.trace() # Indicate that /log endpoint should be traced
+        def log():
+                request_span = tracer.get_span(request)
+                # Extract the span information for request object.
+                with jaeger_tracer.start_span("python webserver internal span of log method",
```
Comment (yurishkuro) [inline example/server.py]: personally I would not use such span name even in the example, because it might make people think this is a good practice.

File: example/README.md (hunk 14 above)
Comment (ror6ax) [reply example/README.md]: It's the way Jaeger manual presents it. I don't want to confuse people who may not be into details of Docker. In any case, ports do not hurt anyone. 
Agree?
